{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2c5b9cfe-eb62-4823-8008-1e9ecac2d930",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.9/site-packages/geopandas/_compat.py:124: UserWarning: The Shapely GEOS version (3.11.1-CAPI-1.17.1) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import dask.dataframe as dd\n",
    "import dask_geopandas as dg\n",
    "import geopandas as gpd\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from segment_speed_utils.project_vars import (SEGMENT_GCS, \n",
    "                                              CONFIG_PATH, \n",
    "                                              PROJECT_CRS\n",
    "                                             )\n",
    "from segment_speed_utils import (helpers, wrangle_shapes, \n",
    "                                 segment_calcs)\n",
    "from shared_utils import rt_dates\n",
    "from shared_utils.geography_utils import WGS84\n",
    "\n",
    "import test_split\n",
    "\n",
    "analysis_date = rt_dates.DATES[\"jul2023\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed5ebb6b-934d-4451-a108-82895f5b0661",
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload\n",
    "\n",
    "dict_inputs = helpers.get_parameters(CONFIG_PATH, \"stop_segments\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7ab12a8-9fef-4e5a-8be3-ad18992d685b",
   "metadata": {},
   "outputs": [],
   "source": [
    "USABLE_VP = dict_inputs[\"stage1\"]\n",
    "INPUT_FILE = dict_inputs[\"stage3\"]\n",
    "SEGMENT_FILE = dict_inputs[\"segments_file\"]\n",
    "SEGMENT_IDENTIFIER_COLS = dict_inputs[\"segment_identifier_cols\"]\n",
    "GROUPING_COL = dict_inputs[\"grouping_col\"]\n",
    "TIMESTAMP_COL = dict_inputs[\"timestamp_col\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29ef4e7b-82d2-427a-9122-f4ee1c34baa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "usable_vp = dd.read_parquet(\n",
    "    f\"{SEGMENT_GCS}{USABLE_VP}_{analysis_date}\",\n",
    "    columns = [\"trip_instance_key\", \"vp_idx\", TIMESTAMP_COL, \"x\", \"y\"]\n",
    ")\n",
    "vp_idx_bounds = test_split.get_usable_vp_bounds_by_trip(usable_vp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "86983844-a321-4d86-863a-3297bd3f7dd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask import delayed, compute\n",
    "\n",
    "# Start from pared down vp\n",
    "df = delayed(pd.read_parquet)(\n",
    "    f\"{SEGMENT_GCS}vp_pare_down/{INPUT_FILE}_all_{analysis_date}\",\n",
    "    columns = SEGMENT_IDENTIFIER_COLS + [\"trip_instance_key\", \"vp_idx\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "784086ef-5e92-4c06-ada0-1abe7819de2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_parquet(\n",
    "    f\"{SEGMENT_GCS}vp_pare_down/{INPUT_FILE}_all_{analysis_date}\",\n",
    "    #columns = SEGMENT_IDENTIFIER_COLS + [\"trip_instance_key\", \"vp_idx\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fd0b0494-8ca7-4a7f-8ff9-3f9b4ac4cced",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_wide(\n",
    "    df: pd.DataFrame, \n",
    "    segment_identifier_cols: list,\n",
    "    timestamp_col: str\n",
    ") -> pd.DataFrame:\n",
    "    \n",
    "    group_cols = [\"trip_instance_key\"] + segment_identifier_cols\n",
    "    \n",
    "    # Sort and make sure we get a pairing of point 1 and 2 within segment\n",
    "    df = df.sort_values(group_cols + [\"vp_idx\"]).reset_index(drop=True)\n",
    "    \n",
    "    df = dd.from_pandas(df, npartitions=50, sort=True)\n",
    "\n",
    "    df = df.assign(\n",
    "        obs = df.groupby(group_cols, \n",
    "                         observed=True, group_keys=False\n",
    "                        ).cumcount() + 1\n",
    "    )\n",
    "    \n",
    "    gdf = dg.from_dask_dataframe(\n",
    "        df, \n",
    "        geometry=dg.points_from_xy(df, \"x\", \"y\")\n",
    "    ).set_crs(WGS84).to_crs(PROJECT_CRS).drop(columns = [\"x\", \"y\"])\n",
    "    \n",
    "    segments = helpers.import_segments(\n",
    "        SEGMENT_GCS,\n",
    "        f\"{SEGMENT_FILE}_{analysis_date}\", \n",
    "        columns = segment_identifier_cols + [\"geometry\"]\n",
    "    ).dropna(subset=\"geometry\").reset_index(drop=True)\n",
    "    \n",
    "    gddf = dd.merge(\n",
    "        gdf,\n",
    "        segments,\n",
    "        on = segment_identifier_cols,\n",
    "        how = \"inner\"\n",
    "    ).rename(columns = {\n",
    "        \"geometry_x\": \"vp_geometry\",\n",
    "        \"geometry_y\": \"segment_geometry\"\n",
    "    }).set_geometry(\"vp_geometry\")\n",
    "    \n",
    "    shape_meters_series = gddf.map_partitions(\n",
    "        wrangle_shapes.project_point_geom_onto_linestring,\n",
    "        \"segment_geometry\",\n",
    "        \"vp_geometry\",\n",
    "        meta = (\"shape_meters\", \"float\")\n",
    "    )\n",
    "    \n",
    "    gddf[\"shape_meters\"] = shape_meters_series\n",
    "    \n",
    "    ddf = gddf.drop(columns = [\"vp_geometry\", \"segment_geometry\"]).reset_index(drop=True)\n",
    "    '''\n",
    "    ddf = segment_calcs.convert_timestamp_to_seconds(\n",
    "        ddf, [timestamp_col])\n",
    "    \n",
    "    # Must do point 1 against point 2\n",
    "    # Some segments only have 1 point, have no second point to find\n",
    "    point1 = ddf[ddf.obs==1].reset_index(drop=True)\n",
    "    point2 = ddf[ddf.obs==2].reset_index(drop=True)\n",
    "    \n",
    "    ddf_wide = dd.merge(\n",
    "        point1,\n",
    "        point2,\n",
    "        on = group_cols,\n",
    "        how = \"left\"\n",
    "    ).reset_index(drop=True)\n",
    "    '''\n",
    "    return ddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "540db34a-ae25-4fde-916e-0ed14aa3e0f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n",
      "/opt/conda/lib/python3.9/site-packages/shapely/linear.py:90: RuntimeWarning: invalid value encountered in line_locate_point\n",
      "  return lib.line_locate_point(line, other)\n"
     ]
    }
   ],
   "source": [
    "results = make_wide(\n",
    "    df, \n",
    "    SEGMENT_IDENTIFIER_COLS, \n",
    "    TIMESTAMP_COL\n",
    ").compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a222b1f4-f54a-4b18-acab-93282af02d1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vp_idx</th>\n",
       "      <th>trip_instance_key</th>\n",
       "      <th>location_timestamp_local</th>\n",
       "      <th>shape_array_key</th>\n",
       "      <th>stop_sequence</th>\n",
       "      <th>obs</th>\n",
       "      <th>shape_meters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3140711</td>\n",
       "      <td>0000703b8b86da99dbf637e746c45236</td>\n",
       "      <td>2023-07-12 18:14:58</td>\n",
       "      <td>15b81a6fa853534940f7f1c8f7c3a3ba</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3140712</td>\n",
       "      <td>0000703b8b86da99dbf637e746c45236</td>\n",
       "      <td>2023-07-12 18:15:13</td>\n",
       "      <td>15b81a6fa853534940f7f1c8f7c3a3ba</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3140711</td>\n",
       "      <td>0000703b8b86da99dbf637e746c45236</td>\n",
       "      <td>2023-07-12 18:14:58</td>\n",
       "      <td>15b81a6fa853534940f7f1c8f7c3a3ba</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>13.352249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3140719</td>\n",
       "      <td>0000703b8b86da99dbf637e746c45236</td>\n",
       "      <td>2023-07-12 18:17:31</td>\n",
       "      <td>15b81a6fa853534940f7f1c8f7c3a3ba</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1114.127066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3140717</td>\n",
       "      <td>0000703b8b86da99dbf637e746c45236</td>\n",
       "      <td>2023-07-12 18:17:00</td>\n",
       "      <td>15b81a6fa853534940f7f1c8f7c3a3ba</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>7.565720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99194</th>\n",
       "      <td>6808154</td>\n",
       "      <td>ffffe3ff18b68f1b90b97f583d4601ef</td>\n",
       "      <td>2023-07-12 19:06:37</td>\n",
       "      <td>f8348bfb848a94699bf0a8d3c2be02c9</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>1042.154977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99195</th>\n",
       "      <td>6808154</td>\n",
       "      <td>ffffe3ff18b68f1b90b97f583d4601ef</td>\n",
       "      <td>2023-07-12 19:06:37</td>\n",
       "      <td>f8348bfb848a94699bf0a8d3c2be02c9</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>15.118653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99196</th>\n",
       "      <td>6808156</td>\n",
       "      <td>ffffe3ff18b68f1b90b97f583d4601ef</td>\n",
       "      <td>2023-07-12 19:07:17</td>\n",
       "      <td>f8348bfb848a94699bf0a8d3c2be02c9</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>393.326336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99197</th>\n",
       "      <td>6808156</td>\n",
       "      <td>ffffe3ff18b68f1b90b97f583d4601ef</td>\n",
       "      <td>2023-07-12 19:07:17</td>\n",
       "      <td>f8348bfb848a94699bf0a8d3c2be02c9</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99198</th>\n",
       "      <td>6808160</td>\n",
       "      <td>ffffe3ff18b68f1b90b97f583d4601ef</td>\n",
       "      <td>2023-07-12 19:08:38</td>\n",
       "      <td>f8348bfb848a94699bf0a8d3c2be02c9</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>208.493146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4959972 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        vp_idx                 trip_instance_key location_timestamp_local  \\\n",
       "0      3140711  0000703b8b86da99dbf637e746c45236      2023-07-12 18:14:58   \n",
       "1      3140712  0000703b8b86da99dbf637e746c45236      2023-07-12 18:15:13   \n",
       "2      3140711  0000703b8b86da99dbf637e746c45236      2023-07-12 18:14:58   \n",
       "3      3140719  0000703b8b86da99dbf637e746c45236      2023-07-12 18:17:31   \n",
       "4      3140717  0000703b8b86da99dbf637e746c45236      2023-07-12 18:17:00   \n",
       "...        ...                               ...                      ...   \n",
       "99194  6808154  ffffe3ff18b68f1b90b97f583d4601ef      2023-07-12 19:06:37   \n",
       "99195  6808154  ffffe3ff18b68f1b90b97f583d4601ef      2023-07-12 19:06:37   \n",
       "99196  6808156  ffffe3ff18b68f1b90b97f583d4601ef      2023-07-12 19:07:17   \n",
       "99197  6808156  ffffe3ff18b68f1b90b97f583d4601ef      2023-07-12 19:07:17   \n",
       "99198  6808160  ffffe3ff18b68f1b90b97f583d4601ef      2023-07-12 19:08:38   \n",
       "\n",
       "                        shape_array_key  stop_sequence  obs  shape_meters  \n",
       "0      15b81a6fa853534940f7f1c8f7c3a3ba              1    1      0.000000  \n",
       "1      15b81a6fa853534940f7f1c8f7c3a3ba              1    2      0.000000  \n",
       "2      15b81a6fa853534940f7f1c8f7c3a3ba              2    1     13.352249  \n",
       "3      15b81a6fa853534940f7f1c8f7c3a3ba              2    2   1114.127066  \n",
       "4      15b81a6fa853534940f7f1c8f7c3a3ba              3    1      7.565720  \n",
       "...                                 ...            ...  ...           ...  \n",
       "99194  f8348bfb848a94699bf0a8d3c2be02c9             20    2   1042.154977  \n",
       "99195  f8348bfb848a94699bf0a8d3c2be02c9             21    1     15.118653  \n",
       "99196  f8348bfb848a94699bf0a8d3c2be02c9             21    2    393.326336  \n",
       "99197  f8348bfb848a94699bf0a8d3c2be02c9             22    1      0.000000  \n",
       "99198  f8348bfb848a94699bf0a8d3c2be02c9             22    2    208.493146  \n",
       "\n",
       "[4959972 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45664688-422d-4336-b7ac-dcaa399b803d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = delayed(test_split.get_prior_position_on_segment)(\n",
    "    df, \n",
    "    SEGMENT_IDENTIFIER_COLS,\n",
    "    TIMESTAMP_COL\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da871931-d2fa-42dd-8d66-b5bd6c735728",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = delayed(dd.merge)(\n",
    "    df2,\n",
    "    vp_idx_bounds,\n",
    "    on = \"trip_instance_key\",\n",
    "    how = \"inner\"\n",
    ")\n",
    "\n",
    "df3 = df3.assign(\n",
    "    prior_vp_idx = df3.apply(\n",
    "        lambda x: \n",
    "        x.vp_idx + 1 if (x.prior_vp_idx < x.min_vp_idx) and \n",
    "        (x.vp_idx + 1 <= x.max_vp_idx)\n",
    "        else x.prior_vp_idx, \n",
    "        axis=1)\n",
    ").drop(columns = [\"trip_instance_key\", \"min_vp_idx\", \"max_vp_idx\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aedd4e2-c3a4-4b18-8ca5-9234fc10f992",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = compute(df3)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f8c086-4592-4c8f-9e08-cbe43c90373f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def attach_vp_timestamp_location(\n",
    "    df: pd.DataFrame,\n",
    "    usable_vp: dd.DataFrame,\n",
    "    timestamp_col: str\n",
    ") -> gpd.GeoDataFrame:\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    \n",
    "    # Turn the vp_idx we need into gdf\n",
    "    vp_to_keep = np.union1d(df.vp_idx, df.prior_vp_idx).tolist()\n",
    "    usable_vp2 = usable_vp[usable_vp.vp_idx.isin(vp_to_keep)]\n",
    "    \n",
    "    # Merge in the timestamp and x, y coords \n",
    "    usable_gdf = gpd.GeoDataFrame(\n",
    "        usable_vp2,\n",
    "        geometry = gpd.points_from_xy(usable_vp2.x, usable_vp2.y),\n",
    "        crs = WGS84\n",
    "    ).to_crs(PROJECT_CRS).drop(columns = [\"x\", \"y\"])\n",
    "    \n",
    "    \n",
    "    df_with_xy = pd.merge(\n",
    "        usable_gdf,\n",
    "        df,\n",
    "        on = \"vp_idx\",\n",
    "        how = \"inner\"\n",
    "    ).rename(columns = {\"geometry\": \"vp_geometry\"})\n",
    "    \n",
    "    # Merge again to get timestamp and x, y coords of previous point\n",
    "    usable_gdf2 = usable_gdf.rename(\n",
    "        columns = {\n",
    "            \"vp_idx\": \"prior_vp_idx\",\n",
    "            timestamp_col: f\"prior_{timestamp_col}\",\n",
    "            \"geometry\": \"prior_vp_geometry\"\n",
    "        }\n",
    "    ).drop(columns = \"trip_instance_key\")\n",
    "    \n",
    "    df_with_prior_xy = dd.merge(\n",
    "        df_with_xy,\n",
    "        usable_gdf2,\n",
    "        on = \"prior_vp_idx\",\n",
    "        how = \"inner\"\n",
    "    )\n",
    "    \n",
    "    return df_with_prior_xy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502f4715-a298-4520-b571-fae4d196cf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from shared_utils.geography_utils import WGS84\n",
    "\n",
    "gdf = delayed(attach_vp_timestamp_location)(\n",
    "    df3,\n",
    "    usable_vp,\n",
    "    TIMESTAMP_COL\n",
    ").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeecaba2-9e58-4954-9660-5337ef02569c",
   "metadata": {},
   "outputs": [],
   "source": [
    "part2 = gdf[gdf.n_vp_seg==2]\n",
    "\n",
    "part2_keep = (part2.groupby([\"trip_instance_key\"] + SEGMENT_IDENTIFIER_COLS,\n",
    "                            observed=True, group_keys=False)\n",
    "              .vp_idx\n",
    "              .max()\n",
    "              .reset_index()\n",
    "             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "150a9219-94c7-44cc-84ac-e9badd88c3e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "part2_pared = delayed(dd.merge)(\n",
    "    part2,\n",
    "    part2_keep,\n",
    "    on = [\"trip_instance_key\", \"vp_idx\"] + SEGMENT_IDENTIFIER_COLS, \n",
    "    how = \"inner\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b7f596-1676-40e8-a185-c0f02ad7c7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "part2_gdf = delayed(test_split.merge_in_segments)(\n",
    "    part2_pared,\n",
    "    SEGMENT_IDENTIFIER_COLS,\n",
    "    GROUPING_COL,\n",
    "    n_vp_seg_value=2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cefe0484-6a54-4b11-a562-5e77ff824b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2_gdf = compute(part2_gdf)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88525881-e849-4a42-b0a3-aa14d9d38f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "gddf = dg.from_geopandas(p2_gdf, npartitions=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "678c1e78-9fe3-4f11-a2f6-4fa7ced6c69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "shape_meters_series = gddf.map_partitions(\n",
    "    wrangle_shapes.project_point_geom_onto_linestring,\n",
    "    \"geometry\",\n",
    "    \"vp_geometry\",\n",
    "    meta = (\"shape_meters\", \"float\")\n",
    ")\n",
    "\n",
    "prior_shape_meters_series = gddf.map_partitions(\n",
    "    wrangle_shapes.project_point_geom_onto_linestring,\n",
    "    \"geometry\",\n",
    "    \"prior_vp_geometry\",\n",
    "    meta = (\"prior_shape_meters\", \"float\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "351cc3d5-ade0-47a3-ba30-aa8e24b4c9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gddf[\"difference_shape_meters\"] = abs(\n",
    "        shape_meters_series - prior_shape_meters_series)\n",
    "gddf[\"straight_distance\"] = gddf.vp_geometry.distance(gddf.prior_vp_geometry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f943943-e0bc-4526-b9ab-e8963eeec7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "scaling_factor = 1.75\n",
    "min_meters_elapsed = 100\n",
    "\n",
    "# It's not only that difference_shape_meters==0 is wrong,\n",
    "# Separate out these and try again with full shape \n",
    "gddf = gddf.assign(\n",
    "    meters_elapsed = gddf.apply(\n",
    "        lambda x: x.straight_distance if (\n",
    "            x.difference_shape_meters <= min_meters_elapsed or \n",
    "            x.difference_shape_meters >= x.straight_distance*scaling_factor\n",
    "        ) else x.difference_shape_meters, \n",
    "        axis=1, \n",
    "        meta = (\"meters_elapsed\", \"float\")\n",
    "    ),\n",
    ")\n",
    "\n",
    "gddf = segment_calcs.convert_timestamp_to_seconds(\n",
    "    gddf, [TIMESTAMP_COL, f\"prior_{TIMESTAMP_COL}\"])\n",
    "\n",
    "gddf = gddf.assign(\n",
    "    sec_elapsed = (gddf[f\"{TIMESTAMP_COL}_sec\"] - \n",
    "                      gddf[f\"prior_{TIMESTAMP_COL}_sec\"]).abs()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b10cbeb9-558e-459b-93d3-becb118ba7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ddf = gddf.drop(columns = [\"prior_vp_geometry\", \n",
    "                           \"vp_geometry\", \"geometry\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d54fb3-d5dc-4a98-a304-9b5004e022b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ddf = ddf.repartition(npartitions=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df1d5370-3021-45fb-a6ef-7eb408185d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = ddf.compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4c9fdf6-55b8-479a-94ae-ab36a4343171",
   "metadata": {},
   "outputs": [],
   "source": [
    "from shared_utils.rt_utils import MPH_PER_MPS\n",
    "\n",
    "results = results.assign(\n",
    "    speed_mph = (results.meters_elapsed.divide(results.sec_elapsed) * \n",
    "                 MPH_PER_MPS)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "433e0382-461e-4970-9d88-37e07a40efaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "results[results.speed_mph < 5].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be27c87b-0de2-472e-9ea0-cc42da1d4256",
   "metadata": {},
   "outputs": [],
   "source": [
    "ok_results = results[results.speed_mph >= 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e177393a-86c0-4c74-824c-fe9548b461b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_results = results[(results.speed_mph >= 0) & (results.speed_mph <5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab46336-10a3-4a25-86d5-6f9343f3175f",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(ok_results), len(bad_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6390361b-bfdd-44f3-833b-12b291343bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "989bdfe8-8dde-4457-b4d9-ca110ecfabf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "part1 = gdf[gdf.n_vp_seg==1]\n",
    "\n",
    "p1 = compute(part1)[0]\n",
    "\n",
    "p1.to_parquet(\"one_vp_in_seg.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1415ef3d-7a88-48ca-ad9c-a20a00618d61",
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_shapes = bad_results.shape_array_key.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce58f30c-54b7-4ba1-854e-fcd9d0314012",
   "metadata": {},
   "outputs": [],
   "source": [
    "ok_results.to_parquet(\"ok_results.parquet\")\n",
    "bad_results.to_parquet(\"bad_results.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77a87e87-77cf-4d90-bdba-a9f6eaaf6433",
   "metadata": {},
   "source": [
    "For speeds that are unusually low, it is now confirmed it's not just `loop_or_inlining`. It happens on `loop_or_inlining==0` too, so it's good to have moved the sjoin postprocessing to all shapes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ae778cb-00df-4c31-9161-c969fa27565a",
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = gpd.read_parquet(\n",
    "    f\"{SEGMENT_GCS}stop_segments_{analysis_date}.parquet\",\n",
    "    filters = [[(GROUPING_COL, \"in\", bad_shapes)]],\n",
    "    columns = SEGMENT_IDENTIFIER_COLS + [\"loop_or_inlining\", \"geometry\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53aab44a-b49f-4658-a79f-cf4cdf1e1383",
   "metadata": {},
   "outputs": [],
   "source": [
    "segments.loop_or_inlining.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bda0d4da-96f9-4833-95f2-c02f992a63c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
